File: lib/test/t_KPermutationsDistribution_std.cxx
//                                               -*- C++ -*-
/**
 *  @file  t_KPermutationsDistribution_std.cxx
 *  @brief The test file of class KPermutationsDistribution for standard methods
 *
 *  Copyright (C) 2005-2013 EDF-EADS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  @author schueller
 *  @date   2012-07-16 15:59:45 +0200 (Mon, 16 Jul 2012)
 */
#include "OT.hxx"
#include "OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

class TestObject : public KPermutationsDistribution
{
public:
  explicit TestObject() : KPermutationsDistribution(5, 12) {}
  explicit TestObject(const String & name) : KPermutationsDistribution(5, 12)
  {
    setName(name);
  }
  virtual ~TestObject() {}
};


int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();
  try
    {
      // Test basic functionnalities
      checkClassWithClassName<TestObject>();

      // Test some extra functionnalities
      checkNameFeature<TestObject>();

      // Instanciate one distribution object
      KPermutationsDistribution distribution(5, 12);
      fullprint << "Distribution " << distribution << std::endl;
      std::cout << "Distribution " << distribution << std::endl;

      // Is this distribution elliptical ?
      fullprint << "Elliptical = " << (distribution.isElliptical() ? "true" : "false") << std::endl;

      // Is this distribution continuous ?
      fullprint << "Continuous = " << (distribution.isContinuous() ? "true" : "false") << std::endl;

      // Test for realization of distribution
      NumericalPoint oneRealization = distribution.getRealization();
      fullprint << "oneRealization=" << oneRealization << std::endl;

      // Test for sampling
      UnsignedLong size = 10000;
      NumericalSample oneSample = distribution.getSample( size );
      fullprint << "oneSample first=" << oneSample[0] << " last=" << oneSample[size - 1] << std::endl;
      fullprint << "mean=" << oneSample.computeMean() << std::endl;
      fullprint << "covariance=" << oneSample.computeCovariance() << std::endl;
      // Define a point
      NumericalPoint point( distribution.getDimension(), 1.0 );
      fullprint << "Point= " << point << std::endl;

      // Show PDF and CDF of point
      NumericalScalar LPDF = distribution.computeLogPDF( point );
      fullprint << "log pdf=" << LPDF << std::endl;
      NumericalScalar PDF = distribution.computePDF( point );
      fullprint << "pdf     =" << PDF << std::endl;
      NumericalScalar CDF = distribution.computeCDF( point );
      fullprint << "cdf=" << CDF << std::endl;
      NumericalScalar CCDF = distribution.computeComplementaryCDF( point );
      fullprint << "ccdf=" << CCDF << std::endl;
      NumericalPoint quantile = distribution.computeQuantile( 0.95 );
      fullprint << "quantile=" << quantile << std::endl;
      fullprint << "cdf(quantile)=" << distribution.computeCDF(quantile) << std::endl;
      NumericalPoint mean = distribution.getMean();
      fullprint << "mean=" << mean << std::endl;
      CovarianceMatrix covariance = distribution.getCovariance();
      fullprint << "covariance=" << covariance << std::endl;
      KPermutationsDistribution::NumericalPointWithDescriptionCollection parameters = distribution.getParametersCollection();
      fullprint << "parameters=" << parameters << std::endl;
    }
  catch (TestFailed & ex)
    {
      std::cerr << ex << std::endl;
      return ExitCode::Error;
    }


  return ExitCode::Success;
}

File: python/test/t_KPermutationsDistribution_std.py
#! /usr/bin/env python

from openturns import *

TESTPREAMBLE()
RandomGenerator().SetSeed(0)

try :
      # Instanciate one distribution object
      distribution = KPermutationsDistribution(5, 12)
      print "Distribution ", distribution

      # Is this distribution elliptical ?
      print "Elliptical = ", distribution.isElliptical()

      # Is this distribution continuous ?
      print "Continuous = ", distribution.isContinuous()

      # Test for realization of distribution
      oneRealization = distribution.getRealization()
      print "oneRealization=", oneRealization

      # Test for sampling
      size = 10000
      oneSample = distribution.getSample( size )
      print "oneSample first=", oneSample[0], " last=", oneSample[size - 1]
      print "mean=", oneSample.computeMean()
      print "covariance=", oneSample.computeCovariance()
      # Define a point
      point = NumericalPoint( distribution.getDimension(), 4.0 )
      print "Point= ", point

      # Show PDF and CDF of point
      LPDF = distribution.computeLogPDF( point )
      print "log pdf=", LPDF
      PDF = distribution.computePDF( point )
      print "pdf     =", PDF
      CDF = distribution.computeCDF( point )
      print "cdf=", CDF
      CCDF = distribution.computeComplementaryCDF( point )
      print "ccdf=", CCDF
      quantile = distribution.computeQuantile( 0.95 )
      print "quantile=", quantile
      print "cdf(quantile)=", distribution.computeCDF(quantile)
      mean = distribution.getMean()
      print "mean=", mean
      covariance = distribution.getCovariance()
      print "covariance=", covariance
      parameters = distribution.getParametersCollection()
      print "parameters=", parameters

except :
    import sys
    print "t_KPermutationsDistribution_std.py", sys.exc_type, sys.exc_value
